include/deal.II/base/work_stream.h

```diff
@@ -677,17 +682,109 @@ namespace WorkStream
+          Worker                                   worker,
+          Copier                                   copier,
+          const ScratchData &                      sample_scratch_data,
+          const CopyData &                         sample_copy_data,
+          const unsigned int                       queue_length,
+          const unsigned int                       chunk_size)
+      {
+        // create the three stages of the pipeline
+        IteratorRangeToItemStream<Iterator, ScratchData, CopyData>
+          iterator_range_to_item_stream(begin,
+                                        end,
+                                        queue_length,
+                                        chunk_size,
+                                        sample_scratch_data,
+                                        sample_copy_data);
+
+        TBBWorker<Iterator, ScratchData, CopyData> worker_filter(worker);
+        TBBCopier<Iterator, ScratchData, CopyData> copier_filter(copier);
+
+        // now create a pipeline from these stages
+        tbb::pipeline assembly_line;
+        assembly_line.add_filter(iterator_range_to_item_stream);
+        assembly_line.add_filter(worker_filter);
+        assembly_line.add_filter(copier_filter);
+
+        // and run it
+        assembly_line.run(queue_length);
+
+        assembly_line.clear();
+      }
+
+    }    // namespace tbb_no_coloring
+#  endif // DEAL_II_WITH_TBB
+
+
+    /**
+     * A reference implementation without using multithreading to be used if we
+     * don't have multithreading support or if the user requests to run things
+     * sequentially. This is more efficient than using TBB or taskflow if we
+     * only have a single thread.
+     */
+    namespace sequential
+    {
+      /**
+       * Sequential version without colors.
+       */
+      template <typename Worker,
+                typename Copier,
+                typename Iterator,
+                typename ScratchData,
+                typename CopyData>
+      void
+      run(const Iterator &                         begin,
+          const typename identity<Iterator>::type &end,
+          Worker                                   worker,
+          Copier                                   copier,
+          const ScratchData &                      sample_scratch_data,
+          const CopyData &                         sample_copy_data)
+      {
+        // need to copy the sample since it is marked const
+        ScratchData scratch_data = sample_scratch_data;
+        CopyData    copy_data    = sample_copy_data; // NOLINT
+
+        // Optimization: Check if the functions are not the zero function. To
+        // check zero-ness, create a C++ function out of it:
+        const bool have_worker =
+          (static_cast<const std::function<
+             void(const Iterator &, ScratchData &, CopyData &)> &>(worker)) !=
+          nullptr;
+        const bool have_copier =
```

Comment:
These 3 lines should be the only change in functionality.

Reply:
Is static cast working? Shouldn't it be dynamic?

Reply:
The only thing I did is move it outside the loop, see https://github.com/dealii/dealii/blob/e8f99f7e31c46c7e57ffa66076e6422d72dd4552/include/deal.II/base/work_stream.h#L1035-L1042

Reply:
`static_cast` is correct. If the type of `copier` is already such a std::function object, then the type does nothing. If it is any other function-like object, then it's cast to a temporary variable for which we can check that it's non-empty. The typical example would be if someone passed `nullptr` as argument.